tiledb/sm/array_schema/dimension.cc

```diff
@@ -1306,28 +1374,90 @@ ByteVecValue Dimension::map_from_uint64(
-  value_T += *(const T*)dim->domain().start();
-
   auto dom_start_T = *(const T*)dim->domain().start();
   auto dom_end_T = *(const T*)dim->domain().end();
   double dom_range_T = dom_end_T - dom_start_T;
 
   // Essentially take the largest value in the bucket
-  if (std::is_integral<T>::value) {  // Integers
-    T norm_coord_T =
-        ceil(((value + 1) / (double)max_bucket_val) * dom_range_T - 1);
-    T coord_T = norm_coord_T + dom_start_T;
-    std::memcpy(&ret[0], &coord_T, sizeof(T));
-  } else {  // Floating point types
-    T norm_coord_T = ((value + 1) / (double)max_bucket_val) * dom_range_T;
-    norm_coord_T =
-        std::nextafter(norm_coord_T, std::numeric_limits<T>::lowest());
-    T coord_T = norm_coord_T + dom_start_T;
-    std::memcpy(&ret[0], &coord_T, sizeof(T));
+  T norm_coord_T =
+      ceil(((value + 1) / (double)max_bucket_val) * dom_range_T - 1);
+  T coord_T = norm_coord_T + dom_start_T;
+  std::memcpy(&ret[0], &coord_T, sizeof(T));
+
+  return ret;
+}
+
+template <
+    class T,
+    typename std::enable_if<std::is_floating_point<T>::value>::type*>
+ByteVecValue Dimension::map_from_uint64(
+    const Dimension* dim, uint64_t value, int bits, uint64_t max_bucket_val) {
+  assert(dim != nullptr);
+  assert(!dim->domain().empty());
+  (void)bits;  // Not needed here
+
+  ByteVecValue ret(sizeof(T));
+
+  // For doubles, we divide by 2 to avoid overflows.
+  double div = std::is_same<double, T>::value ? 2 : 1;
+
+  double dom_start = *(const T*)dim->domain().start() / div;
+  double dom_end = *(const T*)dim->domain().end() / div;
+  auto dom_range = dom_end - dom_start;
+
+  /** Here, we need to return the maximum value that will lead to this bucket.
+   * Unfortunately, rounding errors won't allow us to be fully accurate.
+   * So we are going to take a value in the middle of this bucket, and a
+   * value in the middle of the next bucket, and do a bisection search. */
+  double norm_coord_low =
+      (((double)value + 0.5) / (double)max_bucket_val) * dom_range;
+  T coord_low = norm_coord_low + dom_start;
+
+  /** If we are using floats and there are more bits for the hilbert buckets
+   * than bits in the float mantissa (24), the next bucket value will be a
+   * power of two of the bits difference. */
+  double next_bucket_val = 1.5;
+  if (std::is_same<float, T>::value) {
+    if (bits > 24) {
+      next_bucket_val = 1 << (bits - 24);
+    }
   }
 
+  double norm_coord_high =
+      (((double)value + next_bucket_val) / (double)max_bucket_val) * dom_range;
+  T coord_high = norm_coord_high + dom_start;
+
+  /** Here, we always keep coord_low in the current bucket and coord_high in
+   * the next, when low is the value after high, we have found out value.
+   * i is used to garantee convergence, which should be done in 53 iterations
+   * as double representation has 53 bits mantissa. */
+  int i = 0;
+  do {
+    // Calculate the middle value.
+    T mid = (coord_high + coord_low) / 2;
```

Comment:
Looking at the code now, the loop can terminate if `mid` equals either `coord_high` or `coord_low`. When the high and low limits are close enough, averaging them can't create a different representable floating point number. This is essentially the same check that the high limit is the next representable after the low limit, but can be checked with just equality tests.